src/main/cpp/util/Parallel.h

```diff
@@ -0,0 +1,189 @@
+{
+	ConcurrentQueue<unsigned int> thread_id_pool;
+	for (unsigned int i = 0; i < num_threads; i++) {
+		thread_id_pool.Enqueue(i);
+	}
+
+	tbb::task_scheduler_init init(num_threads);
+
+	tbb::parallel_for(
+	    tbb::blocked_range<size_t>(0, values.size()),
+	    [&action, &values, &thread_id_pool](const tbb::blocked_range<size_t>& r) {
+		    auto thread_id = thread_id_pool.Dequeue();
+		    for (size_t i = r.begin(); i != r.end(); i++) {
+			    action(values[i], thread_id);
+		    }
+		    thread_id_pool.Enqueue(thread_id);
+	    });
+}
+
+#elif defined PARALLELIZATION_LIBRARY_STL
+
+/// The name of the parallelization library that is in use.
+const char* const parallelization_library_name = "STL";
+
+/// Tells if a parallelization library is in use.
+const bool using_parallelization_library = true;
+
+template <typename T, typename TAction>
+void parallel_for(std::vector<T>& values, unsigned int num_threads, const TAction& action)
+{
+	if (num_threads <= 1) {
+		// Nothing to parallelize.
+		serial_for(values, action);
+	} else {
+		// Create num_thread threads and divide the workload statically.
+		std::vector<std::thread> thread_pool;
+		size_t workload_per_thread = values.size() / num_threads;
+		size_t values_start_offset = 0;
+		for (unsigned int i = 0; i < num_threads; i++) {
+			size_t next_start_offset = values_start_offset + workload_per_thread;
+			size_t values_end_offset = i == num_threads - 1 ? values.size() : next_start_offset;
+			if (values_end_offset == next_start_offset) {
+				continue;
+			}
+
+			thread_pool.emplace_back([&values, &action, i, values_start_offset, values_end_offset] {
+				for (size_t j = values_start_offset; j < values_end_offset; j++) {
+					action(values[j], i);
+				}
+			});
+		}
+
+		// Wait for the threads to finish.
+		for (auto& thread : thread_pool) {
+			thread.join();
+		}
+	}
+}
+
+#elif defined _OPENMP && !defined PARALLELIZATION_LIBRARY_NONE
+
+/// The name of the parallelization library that is in use.
+const char* const parallelization_library_name = "OpenMP";
+
+/// Tells if a parallelization library is in use.
+const bool using_parallelization_library = true;
+
+template <typename T, typename TAction>
+void parallel_for(std::vector<T>& values, unsigned int num_threads, const TAction& action)
+{
+#pragma omp parallel for num_threads(num_threads) schedule(runtime)
+	for (size_t i = 0; i < values.size(); i++) {
+		const unsigned int thread_id = omp_get_thread_num();
+		action(values[i], thread_id);
+	}
+}
+
+#else
+
+/// The name of the parallelization library that is in use.
+const char* const parallelization_library_name = "serial";
+
+/// Tells if a parallelization library is in use.
+const bool using_parallelization_library = false;
+
+template <typename T, typename TAction>
+void parallel_for(std::vector<T>& values, unsigned int num_threads, const TAction& action)
+{
+	for (size_t i = 0; i < values.size(); i++) {
```

Comment:
Technically a very tiny bit of code duplication: this could just be a call to `serial_for`.